src/plugins/action_change_status/src/changeStatus.ts

```diff
@@ -0,0 +1,86 @@
+import { ActionPluginOutput } from '../../types'
+
+type Status =
+  | 'Draft'
+  | 'Withdrawn'
+  | 'Submitted'
+  | 'Changes Required'
+  | 'Re-submitted'
+  | 'Completed'
+
+type IParameters = {
+  applicationId: number
+  newStatus: Status
+}
+
+module.exports['changeStatus'] = async function (
+  parameters: IParameters,
+  DBConnect: any
+): Promise<ActionPluginOutput> {
+  const { applicationId, newStatus } = parameters
+  const returnObject: ActionPluginOutput = { status: null, error_log: '' }
+  console.log(`Changing the Status of Application ${applicationId}...`)
+
+  try {
+    const currentStatus = await DBConnect.getCurrentStatusHistory(applicationId)
+
+    switch (currentStatus?.status) {
+      case undefined:
+        // Create initial status
+        const currentStageHistoryId = await DBConnect.getCurrentStageHistory(applicationId)
+        if (!currentStageHistoryId) {
+          console.log(
+            "No stage defined for this Application. Can't create a status_history record."
+          )
+          returnObject.status = 'Fail'
+          returnObject.error_log = 'Missing stage_history for Application'
+        } else {
+          const result = await DBConnect.addNewStatusHistory(currentStageHistoryId.id, newStatus)
+          if (result.id) {
+            returnObject.status = 'Success'
+            returnObject.output = { status: newStatus, statusId: result.id }
+            console.log(`New status_history created: ${newStatus}`)
+          } else {
+            returnObject.status = 'Fail'
+            returnObject.error_log = "Couldn't create new application_status_history"
+          }
+        }
+        break
+      case newStatus:
+        // Do nothing
+        console.log(
+          `WARNING: Application ${applicationId} already has status: ${newStatus}. No changes were made.`
+        )
+        returnObject.status = 'Success'
+        returnObject.error_log = 'Status not changed'
+        returnObject.output = { status: newStatus, statusId: currentStatus.id }
+        break
+      default:
+        // Create a new application_status_history record
+        const result = await DBConnect.addNewStatusHistory(
```

Comment:
would prefer an early return of the returnObject and then add new status operation:

i.e
```
if(newStatus === currentStatus?.status)
   ...
  return returnObject
if(!currentStageHistoryId)
  ...
  return returnObject

... do the addition.
```

It's a stylistic comment, so not insisting it's changed to match above, I've suggested this way because of repetition of addition logic.

Reply:
Righto, I've done something like that now and removed the addition repetition.